test/bufconn/bufconn.go

```diff
@@ -95,6 +96,16 @@ func (l *Listener) Dial() (net.Conn, error) {
 	}
 }
 
+// DialContext provides a function usable with the grpc.WithContextDial DialOption
+func (l *Listener) DialContext(ctx context.Context, _ string) (net.Conn, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+	return l.Dial()
```

Comment:
Maybe we should also add a `l.dial(ctx)`, and wait on the ctx in `select`?
@dfawley WDYT?

Reply:
The thing I *really* want is to make sure we have something that matches:

```go
func WithContextDialer(f func(context.Context, string) (net.Conn, error)) DialOption
```

so I can simply do:

```go
l := bufcon.Listen(sz)
grpc.WithContextDialer(l.DialerContext)
```

Any deeper changes to get there, I'm happy to contribute to as you folks see fit, but otherwise agnostic about :)

Reply:
Did the missing context cancel check cause problem for you? What if you just ignore the context? (I'm not suggesting ignoring the ctx, just want to clarify)

I don't remember if we ever call the dialer with a canceled ctx, but I think it can happen in a race.

Reply:
@menghanl

> Did the missing context cancel check cause problem for you?

Nope.  Missing context cancel checking did not cause a problem for me.  

I included it out of habitual paranoia :)

I would concur with the assertion that calling dialer with a canceled ctx is pathological... but we've all been around enough to realize it *will* happen :)  And I consider it a kindness to fail in dramatic obvious ways when something really pathological has taken place :)